lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1345,6 +1387,70 @@ static LogicalResult verifyIfOp(IfOp ifOp) {
   return success();
 }
 
+/// Returns the last EnableOp within the child tree of 'parentSeqOp'. If no
+/// EnableOp was found (e.g. a "calyx.par" operation is present), returns
+/// nullptr.
+static EnableOp getLastEnableOp(SeqOp parent) {
+  auto &lastOp = parent.getBody()->back();
+  if (auto enableOp = dyn_cast<EnableOp>(lastOp))
+    return enableOp;
+  else if (auto seqOp = dyn_cast<SeqOp>(lastOp))
+    return getLastEnableOp(seqOp);
+  return nullptr;
+}
+
+/// Removes common tail enable operations for sequential 'then'/'else'
+/// branches inside an 'if' operation.
+///
+///   if %a with %A {                       if %a with %A {
+///     seq { ... calyx.enable @B }           seq { ... }
+///   else {                          ->    } else {
+///     seq { ... calyx.enable @B }           seq { ... }
+///   }                                     }
+///                                         calyx.enable @B
+static LogicalResult eliminateCommonTailEnable(IfOp ifOp,
+                                               PatternRewriter &rewriter) {
+  // Check if the branches exist.
+  if (!ifOp.thenRegionExists() || !ifOp.elseRegionExists())
+    return failure();
+
+  auto &thenOpStructureOp = ifOp.getThenBody()->front();
+  auto &elseOpStructureOp = ifOp.getElseBody()->front();
+  // TODO(circt/#1861): ParOps have less restrictive conditions.
```

Comment:
The `IfOp` region is going to be re-written quite a bit to account for #1861, so I wouldn't stare too hard.